### ISSUE TITLE
Setting to ignore whitespace-only lines

### ISSUE THREAD

**Issue by andrewrothman** [Microsoft/vscode#49661]
VS Code currently removes all trailing whitespace including on whitespace only lines.

Consider the following JavaScript/Typescript snippet:

```
export default class Example {
    constructor() {
        console.log("called constructor");
    }
    
    doSomething() {
        console.log("I did something.");
    }
}
```

When I come back to this file, I expect to be able to move the cursor between the two methods and add a new one in between without having to hit the tab key, but after running the VS Code formatter, the whitespace on that line (ln. 5) is removed.

Atom supports this under an "Ignore Whitespace Only Lines" setting.

Additionally, there's a VS Code extension which adds this functionality.

[Trailing Whitespace VS Code Plugin](https://marketplace.visualstudio.com/items?itemName=shardulm94.trailing-spaces#user-content-include-empty-lines)

 It'd be really great if it were configurable and/or auto-detected on a per-file basis by the editor, and would really help my workflow.

**Comment by Magenda** [comment]
Is there any solution to this isuue with or without an extension? It drives me crazy for months now. I would like the cursor to be placed respecting indentation in empty lines *after formatting*. Thanks!

**Comment by tarsupin** [comment]
+1 for this request. I know it might seem like a small thing, but having my cursor keep in line (at least having an option to ignore whitespace) would really improve my workflow and quality of experience.

**Comment by andrewrothman** [comment]
@alexandrudima Could you shed some light on this issue?

Is this a technically difficult task? Would you guys be open to this functionality? Would you be able to provide some info on what would be required?

Thanks for the work you're doing on VS Code. It's awesome.

Andrew

**Comment by Gkupce** [comment]
@andrewrothman
I don't think it should be a difficult task. 
It's just adding a config that when the trim code encounters a line that ends in whitespace, it checks it doesn't match the regex `^\s+$`

**Comment by tarsupin** [comment]
@andrewrothman @Gkupce I don't know when this appeared, since I found it by accident when looking for something else, but if you look for "Editor: Trim Auto Whitespace" in the preferences and un-select it, it does appear to have the desired functionality. 

I believe this thread can be closed.

**Comment by HoldYourWaffle** [comment]
That option only prevents automatically inserted whitespace being 'immediately' removed. It doesn't help when you're formatting a file.

**Comment by andrewrothman** [comment]
@tarsupin Thanks for the tip. Looks like there's still some unexpected behavior with formatting as per @HoldYourWaffle's comment. Also, I'm not sure if that setting takes this into account or not, but I'd be interested in removing whitespace at the end of non-empty lines, but preserving whitespace-only lines. I never add extra tabs or spaces after a line of code, but I do use whitespace-only lines to navigate easier while I'm writing code.

**Comment by HoldYourWaffle** [comment]
> Also, I'm not sure if that setting takes this into account or not, but I'd be interested in removing whitespace at the end of non-empty lines, but preserving whitespace-only lines. I never add extra tabs or spaces after a line of code, but I do use whitespace-only lines to navigate easier while I'm writing code.

This is exactly what I'd love to have. As far as I know there sadly isn't a built-in solution for this right now.

**Comment by andrewrothman** [comment]
I'd be interested in implementing this. Does anyone know if there is a process to have the idea reviewed/modified/approved by the VS Code team before writing code? Or if someone could provide some guidance in the code architecture? I think this would require changes to the settings system (replacing that setting with an enum dropdown versus a checkbox) and whatever subsystem takes the current setting into account. Also, can anyone spot any potential issues with other formatters / extensions that might impact this?

**Comment by tombrito** [comment]
https://marketplace.visualstudio.com/items?itemName=shardulm94.trailing-spaces

**Comment by tbrophy** [comment]
```
    "files.trimTrailingWhitespace": true
```

...works in a settings.json file for VS Code:

>Version: 1.41.1
Commit: 26076a4de974ead31f97692a0d32f90d735645c0
Date: 2019-12-18T14:57:51.166Z
Electron: 6.1.5
Chrome: 76.0.3809.146
Node.js: 12.4.0
V8: 7.6.303.31-electron.0
OS: Darwin x64 18.7.0

**Comment by Scatterblak** [comment]
> https://marketplace.visualstudio.com/items?itemName=shardulm94.trailing-spaces

Skip this. This highlights and deletes whitespace at the end of lines -- nothing to do with preserving whitespace-only lines. You can skip the whitespace only lines when firing the process off to delete trailing whitespace, but it doesn't change the inherent editor behavior as far as I can tell re: preserving whitespace lines in general.

**Comment by Scatterblak** [comment]
Is there any movement on this? *Really* driving me crazy.

**Comment by qeireal** [comment]
Now we have `"files.trimTrailingWhitespace": true` config for it.

**Comment by luke-uppx** [comment]
How did you solve?

**Comment by luke-uppx** [comment]
@andrewrothman How did you solve?

**Comment by Scatterblak** [comment]
I wish I knew -- I wrestled with this forever, and it works now, but I
can't remember what I changed. I can send my entire config if it'll help?

W. Sean Harrison
CTO, Principal

Hagimo LLC
p: 615.838.9289
e: sean.harrison@hagimo.com

http://www.hagimo.com

The contents of this email message and any attachments are intended solely
for the addressee(s) and may contain confidential and/or privileged
information and may be legally protected from disclosure. If you are not
the intended recipient of this message or their agent, or if this message
has been addressed to you in error, please immediately alert the sender by
reply email and then delete this message and any attachments. If you are
not the intended recipient, you are hereby notified that any use,
dissemination, copying, or storage of this message or its attachments is
strictly prohibited.

**Comment by luke-uppx** [comment]
> I can send my entire config if it'll help? 

@Scatterblak please, if you want.

**Comment by GrantGryczan** [comment]
I have `files.trimTrailingWhitespace` set to `false`, but when I paste code with trailing whitespace on otherwise blank lines, it removes it. My `editor.defaultFormatter` is set to `null`. How can I fix this? Or is this issue still present?

**Comment by FaultyFunctions** [comment]
How is this still not implemented?

**Comment by Karl-EdwardFPJeanMehu** [comment]
It would be really helpful to have this feature.

**Comment by alexreg** [comment]
Still no news, folks? This seems like a very easy feature to implement.

**Comment by Scatterblak** [comment]
I've tried soooooo many settings - at this point, I can't look up and try
to figure out which one worked, but here's
the relevant portion of my config - someone let me know what it was? :)

{
    "editor.fontSize": 12,
    "editor.minimap.enabled": true,
    "window.zoomLevel": 2,
    "breadcrumbs.enabled": true,
    "editor.quickSuggestionsDelay": 5000,
    "html.format.indentInnerHtml": true,
    "editor.detectIndentation": false,
    "editor.insertSpaces": false,
    "editor.suggestSelection": "first",
    "vsintellicode.modify.editor.suggestSelection":
"automaticallyOverrodeDefaultValue",
    "editor.largeFileOptimizations": false,
    "editor.fontFamily": "Ubuntu Mono",
    "editor.minimap.showSlider": "always",
    "editor.acceptSuggestionOnEnter": "off",
    "sftp.debug": true,
    "sftp.printDebugLog": true,
    "editor.wordWrap": "off",
    "editor.trimAutoWhitespace": false,
    "workbench.editor.enablePreview": false,
    "editor.matchBrackets": "never",
    "editor.highlightActiveIndentGuide": false,
    "editor.renderIndentGuides": false,
    "explorer.confirmDelete": false,
    "java.home": "C:\\Program Files\\AdoptOpenJDK\\jdk-11.0.11.9-hotspot",
    "mssql.connections": [

------------All Redacted--------------

    ],
    "php.validate.executablePath": "M:\\php-8.0.3\\php.exe",
    "cfml.engine.name": "lucee",
    "html.format.indentHandlebars": true,
    "cfml.autoCloseTags.enable": false,
    "editor.autoIndent": "keep",
    "files.exclude": {
        "**/.classpath": true,
        "**/.project": true,
        "**/.settings": true,
        "**/.factorypath": true
    },
    "sonarlint.rules": {
        "java:S106": {
            "level": "off"
        }
    },
    "rsp-ui.enableStartServerOnActivation": [
        {
            "id": "redhat.vscode-community-server-connector",
            "name": "Community Server Connector",
            "startOnActivation": true
        }
    ],
    "redhat.telemetry.enabled": false,
    "json.maxItemsComputed": 50000,
    "security.workspace.trust.untrustedFiles": "open"
}

**Comment by yashshah** [comment]
Any updates on this one?

**Comment by jangxyz** [comment]
Sharing my *working* configuration, gathered up from the comments above, just for any of you who is desperate to read all the comments and found nothing. That was me.

In short: I turned off all vscode native configs, and relied on the trailing-spaces extension.


### My status

* Working on JavaScript/TypeScript files, with Prettier formatter.
* I want VS code to...
  - Trim trailing white spaces *on save*, 
  - but NOT the whitespace-only lines.

Sample code:
```typescript
export default class Example {
    constructor() {
        console.log("called constructor");    
    }                                      // ^ space up to here
     
 // ^ space up to here
    doSomething() {
        console.log("I did something.");
    }
}
```

<details>
<summary>Highlighted screenshot</summary>

This is how it looks in my vscode, with whitespaces highlighted by Trailing Spaces and the default `Editor: Render Whitespace` setting set to `trailing`:

<img width="622" alt="image" src="https://user-images.githubusercontent.com/52015/182008977-5f30cc92-92f0-4d1c-9a3d-722aebef417a.png">

</details>

### 1. Disable VS Code stuff

To prevent from interfering, I **turned off** the following configs:
- `Files: Trim Trailing Whitespace` - with this turned on, both whitespaces are trimmed on save.
- `Editor: Trim Auto Whitespace`

   when turned on, it removes whitespaces that was generated automatically (like starting a new block) if no longer needed (See  [version 1.2 update, May, 2016](https://code.visualstudio.com/updates/May_2016#_trim-auto-whitespace)).

- `Editor: Format On Save` - somewhere, some formatter, some configuration may be doing some stuff I am not aware of. 

   If you know exactly which formatter you are using, on which file type, whether it is enabled, following which config file, whether on both the User and Workspace scope, may be you don't need to turn this off. I didn't. Frankly speaking, I can't figure out which *formatter* my vscode is using, whether it is my own prettier, eslint, vscode's default one, or some specific configuration in vscode, or any combinations of them. Turning it off made it easier. You can turn it on again or tweak it once you have the configurations complete. 

### 2. Trailing Spaces

First visit the [marketplace for Trailing Spaces](https://marketplace.visualstudio.com/items?itemName=shardulm94.trailing-spaces) and install it.

The default behavior of the extension is to highlight the trailing whitespaces, as the above screenshot shows.

Additional configs:
- `Trailing-spaces: Include Empty Lines` - turn this **OFF** so that it is not highlighted. This is the behavior that we are looking for.

- `Trailing-spaces: Trim On Save` - turn this ON to replace the original vscode behavior of "trim on save".

<details>
<summary>Screenshot of how it looks after the configurations applied:</summary>

<img width="615" alt="image" src="https://user-images.githubusercontent.com/52015/182009941-af7a52fb-4a33-4a42-a059-e1a18e9c35df.png">

Note the space characters on line 5 are still there, but it is not highlighted anymore, and thus not being removed by both the Trailing Spaces extension (because it is not targetted any more), and vscode (because we turned it off).

</details>

NOTE

For some reason, it still removes *any* white-spaces if I manually trigger the `Format Document` action. I tried disabling both `JavaScript > Format: Enable` and `TypeScript > Format: Enable` settings, but it still behaves so. Only after I disabled the VSCode Prettier extension, it stopped working (requires reload). This kind of shows that I am not complete, and still have some more configurations left.

**Comment by Peter-Schorn** [comment]
@jangxyz But how do I get the extension to stop highlight the trailing whitespace? I just want the whitespace to be remove.

**Comment by PeppeL-G** [comment]
(I don't really contribute with anything new here; I mostly post just to give the issue some new well deserved attention, in hope of getting this tiny and very annoying bug fixed)

This is clearly a bug IMO. Some of us (many of us? Almost all of us?) want to keep "empty lines" indented.

---

Why should indentation on an "empty line" be removed? I can think of two reasons, but none is good:

1. *To save space*, so one doesn't have to save the indentation on the hard drive and save storage that way, but if you're short of storage on the hard drive, some indentation in some files is not your main problem
2. *To avoid having to repeatedly press space/tabs on lines you won't write any code on*, but with a modern editor supporting auto-indent you no longer need to do this (you only needed to do this in very old/bad code editors that don't support auto-indent)

---

Why should indentation on an "empty line" be kept? Because it's logical. All lines within the same scope should be indented equally much, and as programmers/engineers it's important to be consistent. Why complicate things by introducing the exception "if the line is empty, all indentation on it should be removed"? Better to keep it simple and don't introduce such an unnecessary exception. Or at least make it configurable, and don't force it upon all of us.

**Comment by ssge** [comment]
One more scenario for keeping the indentation on blank lines is that it makes it possible to create python code in the editor and immediately after that to copy/past only that part in the REPL and test it. (this is a bug in the REPL paste but still other editor support this workaround)

**Comment by ssge** [comment]
@Scatterblak I am not sure but I think 
`"editor.autoIndent":` `"keep",`
did it.

**Comment by ssge** [comment]
@Scatterblak thanks for sharing. I think this line did it

`        "editor.trimAutoWhitespace": false,`

and the relevant portion of my `settings.json`

```JSON
    "[python]": {
        "editor.trimAutoWhitespace": false,
    },
```
